backend/security/src/main/kotlin/io/tolgee/security/authorization/ProjectAuthorizationInterceptor.kt

```diff
@@ -0,0 +1,202 @@
+            requiredScopes.map { s -> s.value }
+          )
+        }
+
+        bypassed = true
+      }
+    }
+
+    if (authenticationFacade.isProjectApiKeyAuth) {
+      val pak = authenticationFacade.projectApiKey
+      // Verify the key matches the project
+      if (project.id != pak.projectId) {
+        logger.debug(
+          "Rejecting access to proj#{} for user#{} via pak#{} - API Key mismatch",
+          project.id,
+          userId,
+          pak.id
+        )
+
+        throw PermissionException()
+      }
+
+      // Validate scopes set on the key
+      requiredScopes?.forEach {
+        if (!pak.scopes.contains(it)) {
+          logger.debug(
+            "Rejecting access to proj#{} for user#{} via pak#{} - Insufficient permissions granted to PAK",
+            project.id,
+            userId,
+            pak.id
+          )
+
+          throw PermissionException(
+            Message.OPERATION_NOT_PERMITTED,
+            requiredScopes.map { s -> s.value }
+          )
+        }
+      }
+    }
+
+    if (bypassed) {
+      logger.info(
+        "Use of admin privileges: user#{} failed local security checks for proj#{} - bypassing for {} {}",
+        userId,
+        project.id,
+        request.method,
+        request.requestURI,
+      )
+    }
+
+    projectHolder.project = project
+    authenticationFacade.authentication.targetProject = project
+    authenticationFacade.authentication.targetOrganization = organizationService.findDto(project.organizationOwnerId)
+    return true
+  }
+
+  private fun getRequiredScopes(request: HttpServletRequest, handler: HandlerMethod): Array<Scope>? {
+    val defaultPerms = AnnotationUtils.getAnnotation(handler.method, UseDefaultPermissions::class.java)
+    val projectPerms = AnnotationUtils.getAnnotation(handler.method, RequiresProjectPermissions::class.java)
+
+    if (defaultPerms == null && projectPerms == null) {
+      // A permission policy MUST be explicitly defined.
+      throw RuntimeException("No permission policy have been set for URI ${request.requestURI}!")
+    }
+
+    if (defaultPerms != null && projectPerms != null) {
+      // Policy doesn't make sense
+      throw RuntimeException(
+        "Both `@UseDefaultPermissions` and `@RequiresProjectPermissions` have been set for this endpoint!"
+      )
+    }
+
+    if (projectPerms?.scopes?.isEmpty() == true) {
+      // No scopes set for RequiresProjectPermissions
+      throw RuntimeException(
+        "`@RequiresProjectPermissions` requires at least one scope. Use `@UseDefaultPermissions` for any scope."
+      )
+    }
+
+    return projectPerms?.scopes
+  }
+
+  override fun getOrder(): Int {
+    return Ordered.HIGHEST_PRECEDENCE
+  }
+
+  companion object {
+    // Excluded routes form filtering (exact):
+    // - /v2/projects
+    // - /v2/projects/
+    // - /v2/projects/with-stats
+    // - /v2/projects/with-stats/
+    val IS_GLOBAL_RE = "^/v2/projects(?:/|/with-stats|/with-stats/)?$".toRegex()
```

Comment:
I sort of did it this way to explicitly avoid having annotations on these endpoints, and make is so there's no way to create a new endpoint without any check because of an accidental copy/paste.

Having an "easy path" to add endpoints to this list weakens the guarantee that endpoints are all protected and one didn't remember to change an annotation.

I'm unsure how to handle this better. Maybe it's acceptable to allow an annotation disabling checks to exist, but it'd be a dangerous weapon that might shoot you in the foot some day